packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1884 @@
+      typeof requestedPermissions !== 'object' ||
+      Array.isArray(requestedPermissions) ||
+      Object.keys(requestedPermissions).length === 0
+    ) {
+      throw invalidParams({ data: { origin, requestedPermissions } });
+    }
+
+    const perms: RequestedPermissions = requestedPermissions;
+
+    for (const methodName of Object.keys(perms)) {
+      const target = (
+        perms[methodName] as unknown as ExtractPermission<
+          PermissionSpecification,
+          CaveatSpecification
+        >
+      ).parentCapability;
+      if (target !== undefined && methodName !== target) {
+        throw invalidParams({ data: { origin, requestedPermissions } });
+      }
+
+      if (!this.getTargetKey(methodName)) {
+        throw methodNotFound({
+          method: methodName,
+          data: { origin, requestedPermissions },
+        });
+      }
+    }
+  }
+
+  /**
+   * Accepts a permissions request created by
+   * {@link PermissionController.requestPermissions}.
+   *
+   * @param request - The permissions request.
+   */
+  async acceptPermissionsRequest(request: PermissionsRequest): Promise<void> {
+    const { id } = request.metadata;
+
+    if (!this.hasApprovalRequest({ id })) {
+      throw new PermissionsRequestNotFoundError(id);
+    }
+
+    if (Object.keys(request.permissions).length === 0) {
+      this._rejectPermissionsRequest(
+        id,
+        invalidParams({
+          message: 'Must request at least one permission.',
+        }),
+      );
+      return;
+    }
+
+    try {
+      this.messagingSystem.call(
+        'ApprovalController:acceptRequest',
+        id,
+        request,
+      );
+    } catch (error) {
+      // If accepting unexpectedly fails, reject the request and re-throw the
+      // error
+      this._rejectPermissionsRequest(id, error);
+      throw error;
+    }
+  }
+
+  /**
+   * Rejects a permissions request created by
+   * {@link PermissionController.requestPermissions}.
+   *
+   * @param id - The id of the request to be rejected.
+   */
+  async rejectPermissionsRequest(id: string): Promise<void> {
+    if (!this.hasApprovalRequest({ id })) {
+      throw new PermissionsRequestNotFoundError(id);
+    }
+
+    this._rejectPermissionsRequest(id, userRejectedRequest());
+  }
+
+  /**
+   * Checks whether the {@link ApprovalController} has a particular permissions
+   * request.
+   *
+   * @see {@link PermissionController.acceptPermissionsRequest} and
+   * {@link PermissionController.rejectPermissionsRequest} for usage.
+   * @param options - The {@link HasApprovalRequest} options.
+   * @returns Whether the specified request exists.
+   */
+  private hasApprovalRequest(options: { id: string }): boolean {
+    // Typecast: There are only some valid parameter combinations, but we
+    // just won't worry about that here.
+    return this.messagingSystem.call(
+      'ApprovalController:hasRequest',
+      // Typecast: Passing just an id is definitely valid, so I don't know
```

Comment:
Which id is this referring to? I'm a bit confused by these two comments. It seems like it's referring to an older version of the code maybe.

Reply:
Oh I see - I missed that the request type only had an id :+1: